### PR TITLE
WL-5302 Track assignment attachment downloads

### PR DESCRIPTION
**Pull request by buckett**
This watches for assignment downloads and records any download a user
does so that we can output a spreadsheet containing details of when the
download was done and then when they submitted.

### CODE REVIEW

File: assignment/assignment-api/api/src/java/org/sakaiproject/assignment/api/model/DownloadEventItem.hbm.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+<hibernate-mapping>
+    <class name="org.sakaiproject.assignment.api.model.DownloadEventItem" table="ASN_DOWNLOAD_ITEM">
+
+        <id name="id" column="ID">
+            <generator class="native">
+                <param name="sequence">ASN_DL_SEQ</param>
+            </generator>
+        </id>
+
+        <property name="attachment" not-null="true" length="255" index="ASN_DL_I_ATT" unique-key="ASN_DL_ATT_USER">
```
Comment (nicholaswilson100) [inline assignment/assignment-api/api/src/java/org/sakaiproject/assignment/api/model/DownloadEventItem.hbm.xml:16]: Where's the 255 length from?

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/AttachmentParser.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    public Details parse(String attachment) {
+        Matcher matcher = pattern.matcher(attachment);
+        if (matcher.matches()) {
+            return new Details(matcher.group("attachment"), matcher.group("siteId"), matcher.group("id"), matcher.group("filename"));
+        }
+        return null;
+    }
+
+    public static class Details {
```
Comment (nicholaswilson100) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/AttachmentParser.java:26]: Does having this as public static mean that anywhere in the code I can change the static Details class?

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/DownloadEventExporter.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    private AttachmentParser parser;
+    // We don't at the moment want to adjust the timezone as we will just use the server timezone
+    private TimeZone timeZone;
+    private Locale locale;
+
+
+    public DownloadEventExporter(DownloadTrackingService service , String... attachments) {
+        this.service = service;
+        this.attachments = attachments;
+        // We want to keep insertion order.
```
Comment (nicholaswilson100) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/DownloadEventExporter.java:35]: Nice comment

File: assignment/assignment-api/api/src/java/org/sakaiproject/assignment/api/model/DownloadEventItem.hbm.xml (hunk 1 above)
Comment (buckett) [reply assignment/assignment-api/api/src/java/org/sakaiproject/assignment/api/model/DownloadEventItem.hbm.xml:16]: Added comment to hbm to explain this.

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/AttachmentParser.java (hunk 2 above)
Comment (buckett) [reply assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/AttachmentParser.java:26]: It's immutable. This is just used to return multiple values (added JavaDoc to explain that)

**Comment by buckett** [comment]
Build failure is because of SVN access, Dave has fixed for now.

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/DownloadEventListener.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            return;
+        }
+        Event event = (Event) arg;
+        if (ContentHostingService.EVENT_RESOURCE_READ.equals(event.getEvent())) {
+            String resource = event.getResource();
+            AttachmentParser.Details detail = parser.parse(resource);
+            if (detail != null) {
+                String userId = event.getUserId();
+                downloadTrackingService.saveDownload(userId, detail.getAttachment(), event.getEventTime());
+                log.debug("Tracked event to download {} in Site {} by {}", detail.getFilename(), detail.getSiteId(), userId);
```
Comment (nicholaswilson100) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/DownloadEventListener.java:56]: What's the logging requirement here? Is it troubleshooting or info?

Comment (buckett) [reply assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/DownloadEventListener.java:56]: Troubleshooting.

File: assignment/assignment-impl/pack/src/webapp/WEB-INF/components.xml [hunk 5]
```diff
@@ -28,20 +28,21 @@
  		<property name="securityService"><ref bean="org.sakaiproject.authz.api.SecurityService"/></property>
  		<property name="developerHelperService"><ref bean="org.sakaiproject.entitybroker.DeveloperHelperService"/></property>
```
Comment (nicholaswilson100) [inline assignment/assignment-impl/pack/src/webapp/WEB-INF/components.xml:29]: You've got quite a lot non-changes in this file

Comment (nicholaswilson100) [reply assignment/assignment-impl/pack/src/webapp/WEB-INF/components.xml:29]: Maybe not...

File: assignment/assignment-tool/tool/src/webapp/js/assignments.js [hunk 6]
```diff
@@ -723,6 +723,10 @@ ASN.invokeDownloadUrl = function(accessPointUrl, actionString, alertMessage, par
     {
         extraInfoArray[extraInfoArray.length]="feedbackComments=true";
     }
+    if (document.getElementById('downloadTimes') && document.getElementById('downloadTimes').checked)
```
Comment (nicholaswilson100) [inline assignment/assignment-tool/tool/src/webapp/js/assignments.js:726]: Is this going back to Sakai?  What's the current state of play of Sakai and us?

**Comment by lardo-de-arnaud** [comment]
Not going back to sakai. Assignments has been rewritten  since sakai 11

**Comment by nicholaswilson100** [comment]
Cool

**Pull request merged**